### ISSUE TITLE
Find a place for `/pkg`

### ISSUE THREAD

**Issue by vieux** [moby/moby#32989]
In an effort of splitting Moby and Docker, we are reorganizing the repo to avoid confusion (see https://github.com/moby/moby/issues/32871)

We decided  to not move some of the packages at first, so we won't  break projects depending on those.

We need to find a place to put the `pkg/` directory.

Maybe a separate repository ?

**Comment by dnephin** [comment]
Here's my proposal for dealing with `pkg/`:

## Summary

`github.com/moby/moby/pkg` should be split up as following:
1. packages that only have a single consumer should be moved to a sub-package of that consumer
2. packages with heavy coupling should be grouped together and moved to a new repo
3. packages that have poor cohesion should be split up and moved to one of the above locations (new repo or single consumer)
4. in a few rare cases parts of very small packages should be duplicated between the two consumers

The full list of moves is rather long, so I've captured it in this gist: [full pkg moves proposal](https://gist.github.com/dnephin/35dc10f6b6b7017f058a71908b301d38). This gist includes the full list of `pkg` dependencies and consumers, as well as the script I used to generate the list.

### New repos

The following new repos would be created. These are the github urls, but are not necessarily the "import urls". We will likely use our own urls with a service like http://labix.org/gopkg.in. These repo names are purely a suggestion and can easily be changed.

* github.com/docker/plugins
* github.com/moby/go-archive
* github.com/moby/go-fs-utils
* github.com/moby/go-http-utils
* github.com/moby/go-ioutils
* github.com/moby/go-locker
* github.com/moby/go-mount
* github.com/moby/go-random
* github.com/moby/go-system
* github.com/moby/go-testutil

Many of these would be created by combining multiple `pkg/` packages.

### Others

16 packages will be moved to their single consumer.
4 packages need to be split up before they are moved.
3 packages are small enough they can be copied to where they are used

**Comment by unclejack** [comment]
@dnephin What would be the purpose of having the code split up into so many repositories? There are some other projects out there which make the life of a potential contributor a nightmare. The more processes and procedures are put in place to make a single contribution, the less likely it is for someone to go through the process to make that contribution.

**Comment by dnephin** [comment]
The general idea is that each of these new repos are functionally very different and are not called from the same components. Any `pkg`s with high coupling were merged into a single package to minimize the overhead of contributing. We're going from 54 `./pkg` packages down to 10 repos.

It should be extremely unlikely that any contribution to a moby/moby component would need to modify more than one of these. In many cases not a single one of these packages would be modified. If you see any cases where these packages would have to be modified together, please do let me know, and I can update my proposal.

My rational for not moving everything into a single "random-pkg-stuff" repo is that it would actually be considerably worse for contributors. If multiple PRs require modification to the same repo it is very difficult to coordinate the vendoring of those changes. If we had every pkg in a single repo then the frequency of conflicting changes would be much higher.

A single repo of random utilities is also difficult to describe, and has no clear goal or purpose. This makes it hard to find owners for the entire repo, and results in quality issues over time.

I think we have to find a good balance between too many repos, and too few. I think the best way to do that is to look at package dependencies and consumers, and group things together based on their coupling to other packages. That is what I've done here.

**Comment by cpuguy83** [comment]
`system`, `mount`, and `fs` seem fairly related.

Maybe instead of `github.com/moby/go-locker` maybe a more generic repo for concurrency primitives?

**Comment by cpuguy83** [comment]
btw, I wrote locker and always hated the name, just no one suggested a different one.

**Comment by dnephin** [comment]
Ya `go-locker` is the smallest of them. I couldn't find any other packages for concurrency primitives. I guess over time we could add more, so renaming it to something more general sgtm.

Combining system, mount, and fs could work, although there are also dependencies between fs and archive I think. 

This proposal was just an approximation, as the work is done to actually split it out we might discover a better fit, but I think the general plan remains the same.

**Comment by stevvooe** [comment]
I would suggest that we favor using the `internal` package rather than assuming broad reuse of the packages within docker. Taking a package from being useful in Docker to broad use cases takes a lot of time and consideration. Often, the goals of making an independent package and separating docker functionality might be at odds.

A great example is [`pkg/httputils`](https://github.com/moby/moby/tree/master/pkg/httputils). There are really only a few exports that are just utility functions that are specific to the way docker uses them.

Dare I invoke this relationship, but I fear this effort will result in a serious case of leftpad-itis.

If we do find that a package is broadly useful, we should create a process for identifying those candidates and eliminate other options before going this route.

**Comment by dnephin** [comment]
The goal of this effort is support breaking out components. Any `pkg` that is only used by a single component could be moved to an `internal`, or some other subpackage. I've categorized those as "16 packages will be moved to their single consumer.". I don't think these are really "internal" worthy, but it could work if we really want to hide them.

For the other 38 `pkg` internal will not work. They need to be used by multiple components. 

I agree many of these packages will have a suboptimal interface initially. 

Trying to refactor everything right now will greatly delay the work to split out components. Instead of trying to fix everything now we can improve the interface of these packages over time.

That said, I don't think we want to rush this and create 10 new repos right away. This is just a roadmap for when we hit shared components. Maybe as we extract things we'll realize that we don't need some of these repos. 

Since you mentioned `pkg/httputils`, [you'll see from my proposal](https://gist.github.com/dnephin/35dc10f6b6b7017f058a71908b301d38#file-_-proposal-L53-L56) that I have suggested splitting it up into the parts that are only used by one consumer, and the parts that are shared with multiple packages.

The `go-httputils` repo will need to be shared between whatever component provides the http api for backwards compatibility and the docker/cli repo, so we know there will be at least two consumers.

**Comment by stevvooe** [comment]
Creating components _is_ the process of finding the correct interfaces. Splitting these out into packages without refactoring is going to create unforeseen problems.

If we can, it would be best to eliminate as many of these packages as possible. If we do find there are real shared patterns, we can propose and design the package with respect to requirements.

Either way, one less package for you to worry about: https://github.com/moby/moby/pull/33097.

**Comment by dnephin** [comment]
> Creating components is the process of finding the correct interfaces

Yes the interface **for the component**, not the interface for every library used by a component. 

> If we can, it would be best to eliminate as many of these packages as possible.

Yes, that is part of the plan.

> Either way, one less package for you to worry about: #33097.

Not exactly. As you can see in the plan `go-random` was going to be created from `pkg/random` and `pkg/stringid` (and possibly `pkg/namegenerator`). So this change is moving the code in the same direction as the plan, but we still need a repo for `pkg/stringid`, which is used in what will becoming a few different components.

**Comment by dnephin** [comment]
Adding some more details based on a discussion during slack standup.

This proposal is only a guideline to follow as we hit `pkg` that are needed by components. We should prioritize the split out of components first, and address `pkg` libraries as part of each component extraction.

Before creating any new repos the following steps should be performed:
1. check for functions with single callers and move them to the consuming package or a subpackage of it
2. if a function is sufficiently small, and has a small number of consumers, duplicate it
3. finally, if 1 and 2 are not appropriate, either create a proposal for a new repo with a clear purpose, or add it to an existing repo if the functionality matches the purpose of the repo

Given these steps, the following work can be done at any time:
* 16 packages moved to their single consumer
* 4 packages split up before they are moved
* 3 packages can be copied to where they are used

These packages and operations are listed in [the full proposal](https://gist.github.com/dnephin/35dc10f6b6b7017f058a71908b301d38#file-_-proposal)

**Comment by stevvooe** [comment]
@dnephin Consider moving mount/mountinfo to containerd dependencies. We can maintain them there and that won't require another package. Please see https://github.com/containerd/containerd/pull/813.

**Comment by dnephin** [comment]
Perfect, that one was an outlier so moving it to containerd would be great. I've updated the proposal and the summary.

**Comment by vdemeester** [comment]
`pkg/devicemapper` is going to be a little bit a pain to move as it seems to be used a bit in some plugins and stuff (https://github.com/search?utf8=%E2%9C%93&q=%22github.com%2Fdocker%2Fdocker%2Fpkg%2Fdevicemapper%22&type=Code)

**Comment by stevvooe** [comment]
@vdemeester It looks like most of those references are forks of docker or vendored dependencies. Are there any that have direct references?
